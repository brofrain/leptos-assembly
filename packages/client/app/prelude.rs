pub use leptos::{
    create_action,
    event_target_value,
    leptos_dom::*,
    on_cleanup,
    provide_context,
    signal_prelude::*,
    spawn_local,
    update,
    use_context,
    view,
    with,
    Action,
    Callback,
    Children,
    DynAttrs,
    Effect,
    IntoView,
    Params,
    Resource,
    ServerFnError,
    Show,
    StoredValue,
    Suspense,
    View,
};
pub use unocss_classes::uno;

pub use crate::{
    app::composables::i18n::{t, use_i18n},
    prelude::*,
};
