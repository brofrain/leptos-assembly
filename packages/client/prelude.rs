pub use cfg_if::*;
pub use derive_getters::*;
pub use serde::*;
pub use time::ext::NumericalDuration;
