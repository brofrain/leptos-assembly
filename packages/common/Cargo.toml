[package]
name = "common"
version.workspace = true
edition.workspace = true
publish = false

[lints]
workspace = true

[lib]
path = "lib.rs"

[dependencies]
common_logger = { path = "logger", optional = true }
common_macros = { path = "macros", optional = true }
common_utils = { path = "utils", optional = true }
cfg-if.workspace = true
derive-getters.workspace = true
derive_more.workspace = true
derive_setters.workspace = true
educe.workspace = true
serde.workspace = true
leptos.workspace = true
log.workspace = true
time.workspace = true
unocss-classes.workspace = true

[features]
logger = ["dep:common_logger"]
macros = ["dep:common_macros"]
utils = ["dep:common_utils"]
csr = ["common_logger/csr"]
ssr = ["common_logger/ssr"]
default = ["macros"]

[package.metadata.cargo-all-features]
skip_feature_sets = [["csr", "ssr"]]
