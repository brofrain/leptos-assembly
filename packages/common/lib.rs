pub use common_macros::{flatten_mod, flatten_pub_mod};

pub mod macros {
    #[cfg(feature = "client")]
    pub use common_macros::component;
    pub use common_macros::{
        cfg_csr,
        cfg_ssr,
        flatten_mod,
        flatten_pub_mod,
        is_csr,
        is_ssr,
        server, // @kw how to tackle backend?
    };
}

#[cfg(feature = "logger")]
pub mod logger {
    pub use common_logger::*;
}

#[cfg(feature = "utils")]
pub mod utils {
    pub use common_utils::*;
}

pub mod exports {
    cfg_if::cfg_if! {
        if #[cfg(feature = "client")] {
            pub use leptos;
            pub use leptos_i18n;
            pub use unocss_classes;
        }
    }

    pub use derive_getters;
    pub use derive_setters;
    pub use educe;
    pub use log;
    pub use serde;
    pub use time;
}

#[macro_export]
macro_rules! use_macros {
    () => {
        #[macro_use]
        extern crate common;
    };
}

pub mod prelude {
    #[cfg(feature = "client")]
    pub use exports::leptos::{
        self,
        create_action,
        event_target_value,
        leptos_dom::*,
        on_cleanup,
        provide_context,
        signal_prelude::*,
        spawn_local,
        update,
        use_context,
        view,
        with,
        Action,
        Callback,
        Children,
        DynAttrs,
        Effect,
        IntoView,
        Params,
        Resource,
        ServerFnError,
        Show,
        StoredValue,
        Suspense,
        View,
    };
    #[cfg(feature = "client")]
    pub use exports::leptos_i18n::{self, t};
    #[cfg(feature = "client")]
    pub use exports::unocss_classes::{self, uno};
    pub use exports::{
        derive_getters::Getters,
        derive_setters::Setters,
        educe::Educe,
        log,
        serde::{self, Deserialize, Serialize},
    };
    #[cfg(feature = "client")]
    pub use macros::component;
    pub use macros::{flatten_mod, flatten_pub_mod};

    use super::{exports, macros};
}
