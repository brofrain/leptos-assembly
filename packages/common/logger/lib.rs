#![feature(lazy_cell)]

use cfg_if::cfg_if;
use common_macros::{cfg_csr, cfg_ssr};
use fern::Dispatch;

#[cfg(feature = "ssr")]
fn build_server_dispatch(dispatch: Dispatch) -> Dispatch {
    use owo_colors::OwoColorize;
    use time::{macros::format_description, OffsetDateTime};

    let time_format =
        format_description!("[year]-[month]-[day] [hour]:[minute]:[second]");

    dispatch
        .chain(std::io::stdout())
        .format(|out, message, record| {
            let level = record.level();

            macro_rules! bold_string {
                    ($($t:tt)*) => {
                        $($t)*.bold().to_string()
                    };
                }

            let level = match level {
                log::Level::Error => bold_string!(level.red()),
                log::Level::Warn => bold_string!(level.yellow()),
                log::Level::Info => bold_string!(level.blue()),
                log::Level::Debug => bold_string!(level.bright_magenta()),
                log::Level::Trace => bold_string!(level.magenta()),
            };

            let module =
                record.module_path().unwrap_or_else(|| record.target());
            let module = module.green();
            let module = module.dimmed();

            let path = record.file().map(|path| {
                let line = record.line();
                match line {
                    Some(line) => format!("{path}:{line}"),
                    None => path.to_owned(),
                }
            });

            let module = match path {
                Some(path) => {
                    format!("{module} {}", path.dimmed())
                }
                None => module.to_string(),
            };

            let time = OffsetDateTime::now_utc()
                .format(time_format)
                .unwrap_or_else(|_| "Unknown time".to_owned());
            let time = time.cyan();
            let time = time.dimmed();

            out.finish(format_args!("{level} {time} {module} {message}",));
        })
}

pub fn init(app_crate_name: &'static str) {
    let mut dispatch = Dispatch::new();

    // debug / info logs from external crates don't interest us
    dispatch = dispatch.level(log::LevelFilter::Warn);

    cfg_if! {
        if #[cfg(debug_assertions)] {
            dispatch = dispatch.level_for(app_crate_name, log::LevelFilter::Debug);
        } else {
            dispatch = dispatch.level_for(app_crate_name, log::LevelFilter::Info);
        }
    }

    cfg_csr! {
        dispatch = dispatch.chain(fern::Output::call(console_log::log)).format(
            |out, message, _| {
                out.finish(*message);
            },
        );
    }

    cfg_ssr! {
        dispatch = build_server_dispatch(dispatch);
    }

    dispatch.apply().expect("Failed to initialize logger");
}
