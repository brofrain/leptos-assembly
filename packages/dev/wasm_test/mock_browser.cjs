const { GlobalRegistrator } = require("@happy-dom/global-registrator");

exports.mock_browser = () => {
  GlobalRegistrator.register();
  Window = Object;
};
