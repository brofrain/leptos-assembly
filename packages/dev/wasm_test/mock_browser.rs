use std::panic;

use common::vendor::log::{self, Level};

#[allow(unsafe_code)]
mod js {
    use wasm_bindgen::prelude::*;

    #[wasm_bindgen(module = "/wasm_test/mock_browser.cjs")]
    extern "C" {
        pub fn mock_browser();
    }
}

pub fn mock_browser() {
    js::mock_browser();
    _ = console_log::init_with_level(Level::Debug);
    panic::set_hook(Box::new(move |panic_info| {
        let msg = panic_info.to_string();
        log::error!("{msg}");
    }));
}
