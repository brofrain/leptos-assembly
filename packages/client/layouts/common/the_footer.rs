use client_components::{BaseIcon, BaseLink};
use client_composables::i18n;
use client_env::PROJECT_REPOSITORY_URL;
use client_i18n::{t_string, use_i18n};
use client_router::Route;
use exports::client::{icondata as i, prelude::*};
use leptos_use::{use_color_mode, ColorMode, UseColorModeReturn};

#[component]
pub fn TheFooter() -> impl IntoView {
    let i18n = use_i18n();

    let UseColorModeReturn { mode, set_mode, .. } = use_color_mode();

    let toggle_dark = move |_| {
        set_mode(if mode() == ColorMode::Dark {
            ColorMode::Light
        } else {
            ColorMode::Dark
        });
    };

    let toggle_locale = {
        let toggle = i18n::use_toggle_locale_and_push_toast();
        move |_| toggle()
    };

    view! {
        <nav class=uno!["flex", "justify-center", "gap-3", "text-xl", "my-4"]>
            <BaseLink title=Signal::derive(move || t_string!(i18n, nav.link_home)) to=Route::Home>
                <BaseIcon icon=i::AiHomeOutlined/>
            </BaseLink>

            <BaseLink
                title=Signal::derive(move || t_string!(i18n, nav.toggle_dark))
                on:click=toggle_dark
            >
                <BaseIcon icon=i::ChSun class="dark:hidden"/>
                <BaseIcon icon=i::TbMoon class="hidden dark:block"/>
            </BaseLink>

            <BaseLink
                title=Signal::derive(move || t_string!(i18n, nav.toggle_locale))
                on:click=toggle_locale
            >
                <BaseIcon icon=i::IoLanguage/>
            </BaseLink>

            <BaseLink title=Signal::derive(move || t_string!(i18n, nav.link_about)) to=Route::About>
                <BaseIcon icon=i::BsCardText/>
            </BaseLink>

            <BaseLink title="GitHub" to=PROJECT_REPOSITORY_URL>
                <BaseIcon icon=i::AiGithubFilled/>
            </BaseLink>
        </nav>
    }
}
