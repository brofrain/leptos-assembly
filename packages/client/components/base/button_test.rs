use common::prelude::*;
use dev::wasm_test::mock_browser;
use wasm_bindgen_test::wasm_bindgen_test;

use crate::BaseButton;

#[wasm_bindgen_test]
async fn render_children() {
    mock_browser();

    leptos::mount_to_body(
        || view! { <BaseButton>{"Hello, World!"}</BaseButton> },
    );

    let el = document().query_selector("button").unwrap().unwrap();

    let msg = el.text_content().unwrap();
    assert_eq!(&msg, "Hello, World!");
}

#[cfg(not(target_arch = "wasm32"))]
#[test]
fn match_snapshot() {
    use dev::vendor::insta::assert_yaml_snapshot;
    use leptos::ssr::render_to_string;

    let html = render_to_string(|| {
        view! { <BaseButton>{"Hello, World!"}</BaseButton> }.into_view()
    });

    assert_yaml_snapshot!(html);
}
