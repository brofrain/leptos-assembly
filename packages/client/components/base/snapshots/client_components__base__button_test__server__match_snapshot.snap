---
source: packages/client/components/base/button_test.rs
expression: html
---
"<!--hk=0-0-0-1o|leptos-base-button-start--><!--leptos-view|packages-client-components-base-button.rs-11|open--><button   data-hk=\"0-0-0-2\" class=\" inline-block rounded p-x4 p-t1 p-b.5 fw-bold text-accent-interactive-contrast bg-accent-interactive transition-colors cursor-pointer select-none disabled:pointer-events-none\"><!--leptos-view|<BaseButton/>-children|open--><!--hk=0-0-0-3o|leptos-<>-start-->Hello, World!<!--hk=0-0-0-3c|leptos-<>-end--><!--leptos-view|<BaseButton/>-children|close--></button><!--leptos-view|packages-client-components-base-button.rs-11|close--><!--hk=0-0-0-1c|leptos-base-button-end-->"

